isthmus/src/main/java/io/substrait/isthmus/OuterReferenceResolver.java

```diff
@@ -0,0 +1,88 @@
+package io.substrait.isthmus;
+
+import java.util.HashMap;
+import java.util.IdentityHashMap;
+import java.util.Map;
+import org.apache.calcite.rel.RelNode;
+import org.apache.calcite.rel.core.CorrelationId;
+import org.apache.calcite.rel.logical.LogicalFilter;
+import org.apache.calcite.rex.*;
+
+/** Resolve correlated variable and get Depth map for RexFieldAccess */
+public class OuterReferenceResolver extends RelNodeVisitor<RelNode, RuntimeException> {
+
+  private Map<CorrelationId, Integer> nestedDepth;
+  private Map<RexFieldAccess, Integer> fieldAccessDepthMap;
+
+  private RexVisitor rexVisitor = new RexVisitor(this);
+
+  public OuterReferenceResolver() {
+    nestedDepth = new HashMap<>();
+    fieldAccessDepthMap = new IdentityHashMap<>();
+  }
+
+  public int getStepsOut(RexFieldAccess fieldAccess) {
+    return fieldAccessDepthMap.get(fieldAccess);
+  }
+
+  public RelNode apply(RelNode r) {
+    return reverseAccept(r);
+  }
+
+  public Map<RexFieldAccess, Integer> getFieldAccessDepthMap() {
+    return fieldAccessDepthMap;
+  }
+
+  @Override
+  public RelNode visit(LogicalFilter filter) throws RuntimeException {
+    for (CorrelationId id : filter.getVariablesSet()) {
+      if (!nestedDepth.containsKey(id)) {
+        nestedDepth.put(id, 0);
+      }
+    }
+    filter.getCondition().accept(rexVisitor);
+    return super.visit(filter);
+  }
+
+  @Override // FIXME: join / project with expressions (possible contain subquery)
+  public RelNode visitOther(RelNode other) throws RuntimeException {
```

Comment:
Can we do something here so that if we hit those scenarios, we make sure to throw as opposed to ignore or lose the data?

Reply:
I checked Project case (Join case will be a filter condition, which has been handled).  

For Filter's case, `variablesSet`  will be set in SqltoRelConverter's `convertWhere`.  However,  for `Project`, SqlToRelConverter will not set  variablesSet().  It delays to `SubQueryRemoveRule`. Therefore,  the code will throw Unsupported exception if a `Project` expression contain a RexSubquery. 

The `RelNode` API `getVariablesSet` has comment that it is only set for `org.apache.calcite.rel.core.Correlate`. But looks like Calcite's code will set the variablesSet for `Filter`, but not for `Project`. 

I added a UT to verify Exception is thrown for the case of `Project`.